usr/lib/common/mech_openssl.c

```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
     ctx->context = NULL;
     return rv;
 }
+
+static CK_RV calc_rsa_priv_exp(STDLL_TokData_t *tokdata, OBJECT *key_obj,
+                               CK_BYTE *priv_exp, CK_ULONG priv_exp_len)
+{
+    CK_ATTRIBUTE *modulus = NULL, *pub_exp = NULL;
+    CK_ATTRIBUTE *prime1 = NULL, *prime2 = NULL;
+    BN_CTX *bn_ctx;
+    BIGNUM *n, *e, *p, *q, *d;
+    CK_RV rc;
+
+    UNUSED(tokdata);
+
+    bn_ctx = BN_CTX_secure_new();
+    if (bn_ctx == NULL) {
+        TRACE_ERROR("Failed to get CKA_MODULUS\n");
+        return CKR_FUNCTION_FAILED;
+    }
+
+    /* Get modulus a BIGNUM */
+    rc = template_attribute_get_non_empty(key_obj->template, CKA_MODULUS,
+                                          &modulus);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_MODULUS\n");
+        goto done;
+    }
+
+    n = BN_CTX_get(bn_ctx);
+    if (n == NULL ||
+        BN_bin2bn(modulus->pValue, modulus->ulValueLen, n) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for modulus\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(n, BN_FLG_CONSTTIME);
+
+    /* Get public exponent a BIGNUM */
+    rc = template_attribute_get_non_empty(key_obj->template,
+                                          CKA_PUBLIC_EXPONENT, &pub_exp);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_PUBLIC_EXPONENT\n");
+        goto done;
+    }
+
+    e = BN_CTX_get(bn_ctx);
+    if (e == NULL ||
+        BN_bin2bn(pub_exp->pValue, pub_exp->ulValueLen, e) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for public exponent\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(e, BN_FLG_CONSTTIME);
+
+    /* Get prime1 a BIGNUM */
+    rc = template_attribute_get_non_empty(key_obj->template, CKA_PRIME_1,
+                                          &prime1);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_PRIME_1\n");
+        goto done;
+    }
+
+    p = BN_CTX_get(bn_ctx);
+    if (p == NULL ||
+        BN_bin2bn(prime1->pValue, prime1->ulValueLen, p) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for prime1\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(p, BN_FLG_CONSTTIME);
+
+    /* Get prime2 a BIGNUM */
+    rc = template_attribute_get_non_empty(key_obj->template, CKA_PRIME_2,
+                                          &prime2);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_PRIME_2\n");
+        goto done;
+    }
+
+    q = BN_CTX_get(bn_ctx);
+    if (q == NULL ||
+        BN_bin2bn(prime1->pValue, prime1->ulValueLen, q) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for prime2\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(q, BN_FLG_CONSTTIME);
+
+    d = BN_CTX_get(bn_ctx);
+    if (d == NULL) {
+        TRACE_ERROR("BN_CTX_get failed d\n");
```

Comment:
There's a 'd' at the end of the trace string

Reply:
Ah ... should be like "BN_CTX_get failed to get d"

Reply:
will fix